### PR TITLE
[otp_ctrl] Allow partitions to absorb unallocated bits

### PR DESCRIPTION
**Pull request by tjaychen**
Follow on from #7689

### CODE REVIEW

**Comment by tjaychen** [comment]
not going to merge this yet, but figured i would just prep the PR since there were a bunch of things we needed to change on the unittest. 

@timothytrippel quick question. 
In general, what is the best way to handle all these unittest that hard code certain bound checks? My thinking is that these were hardcoded (instead of referencing values in the header) as a means to double check.  But is that true? Or should we used macros for those as well?

**Comment by timothytrippel** [comment]
Great question, I saw something similar in the alert handler DIF and I changed the unit tests to use the generated macros. I am of the belief that the autogenerated macros should be used in the unit tests too, that way the unit tests can adapt to any future HJSON changes without requiring manual updates to the DIF unit tests too (easier to maintain). The tool that's generating the macros from the HJSON should be verified/tested orthogonally to given us guarantees that HJSON values really do match autogenerated macro values. If we require certain macros to remain assigned to specific value, this check should be done in a static_assert(MACRO == <hardcoded value>) in the DIF itself. Hopefully that makes sense!

**Comment by tjaychen** [comment]
> Great question, I saw something similar in the alert handler DIF and I changed the unit tests to use the generated macros. I am of the belief that the autogenerated macros should be used in the unit tests too, that way the unit tests can adapt to any future HJSON changes without requiring manual updates to the DIF unit tests too (easier to maintain). The tool that's generating the macros from the HJSON should be verified/tested orthogonally to given us guarantees that HJSON values really do match autogenerated macro values. If we require certain macros to remain assigned to specific value, this check should be done in a static_assert(MACRO == ) in the DIF itself. Hopefully that makes sense!

that sounds good to me! so it sounds like you would not be against me tweaking the otp_ctrl_unittest a tiny bit to use those header values as well?

**Comment by timothytrippel** [comment]
No, I think that should be encouraged! ... on the condition that if it is really important those current macro values remain constant, such checks are moved to static_asserts() in the DIF itself. (I'm not sure if they are in this case.)

**Comment by tjaychen** [comment]
thoughts on this one anyone? should i go ahead and tweak the unittests to match? then we can work on getting this in?

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 1]
```diff
@@ -483,7 +483,7 @@ TEST_F(DaiProgramTest, Program32) {
   EXPECT_READ32(
       OTP_CTRL_DIRECT_ACCESS_REGWEN_REG_OFFSET,
       {{OTP_CTRL_DIRECT_ACCESS_REGWEN_DIRECT_ACCESS_REGWEN_BIT, true}});
-  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x620);
+  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x6a0);
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:486]: same here

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 2]
```diff
@@ -536,7 +536,7 @@ TEST_F(DaiProgramTest, Unaligned) {
 TEST_F(DaiProgramTest, OutOfRange) {
   // Check that we can't write a digest directly.
   EXPECT_EQ(dif_otp_ctrl_dai_program32(&otp_, kDifOtpCtrlPartitionCreatorSwCfg,
-                                       /*address=*/0x2f8, /*value=*/42),
+                                       /*address=*/0x338, /*value=*/42),
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:539]: same here

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 3]
```diff
@@ -576,7 +576,7 @@ TEST_F(DaiDigestTest, DigestSw) {
   EXPECT_READ32(
       OTP_CTRL_DIRECT_ACCESS_REGWEN_REG_OFFSET,
       {{OTP_CTRL_DIRECT_ACCESS_REGWEN_DIRECT_ACCESS_REGWEN_BIT, true}});
-  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x2f8);
+  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x338);
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:579]: and here

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 4]
```diff
@@ -591,7 +591,7 @@ TEST_F(DaiDigestTest, DigestHw) {
   EXPECT_READ32(
       OTP_CTRL_DIRECT_ACCESS_REGWEN_REG_OFFSET,
       {{OTP_CTRL_DIRECT_ACCESS_REGWEN_DIRECT_ACCESS_REGWEN_BIT, true}});
-  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x600);
+  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x680);
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:594]: and here

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 5]
```diff
@@ -722,7 +722,7 @@ class BlockingIoTest : public OtpTest {
 
 TEST_F(BlockingIoTest, Read) {
   for (int i = 0; i < kWords; ++i) {
-    auto offset = 0x300 + 0x10 + i * sizeof(uint32_t);
+    auto offset = 0x340 + 0x10 + i * sizeof(uint32_t);
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:725]: and here

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 6]
```diff
@@ -753,7 +753,7 @@ TEST_F(BlockingIoTest, OutOfRange) {
                                        0x300, buf.data(), buf.size()),
             kDifOtpCtrlDaiOutOfRange);
   EXPECT_EQ(dif_otp_ctrl_read_blocking(&otp_, kDifOtpCtrlPartitionOwnerSwCfg,
-                                       0x10, buf.data(), 0x2f0),
+                                       0x10, buf.data(), 0x330),
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:756]: and here

File: sw/device/lib/dif/dif_otp_ctrl_unittest.cc [hunk 7]
```diff
@@ -385,7 +385,7 @@ TEST_F(DaiReadTest, Read32) {
   EXPECT_READ32(
       OTP_CTRL_DIRECT_ACCESS_REGWEN_REG_OFFSET,
       {{OTP_CTRL_DIRECT_ACCESS_REGWEN_DIRECT_ACCESS_REGWEN_BIT, true}});
-  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x620);
+  EXPECT_WRITE32(OTP_CTRL_DIRECT_ACCESS_ADDRESS_REG_OFFSET, 0x6a0);
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_otp_ctrl_unittest.cc:388]: can we use the OTP_CTRL_PARAM_MANUF_STATE_OFFSET macro here? (instead of 0x6a0)?

**Review by timothytrippel — COMMENTED** [review]
this sounds logical, however, if the sw partition can change at anytime, maybe adding a static_assert somewhere in the unit test file that will cause the build to break if it ever shrinks from what it is now (assuming everything is in a good working condition now) to act as a canary may be a good fail safe here?

other than the few unit test fixes, may want to get @cfrantz / @mundaym to weigh in here, since they are more familiar with the mask rom than I.

**Review by cindychip — COMMENTED** [review]
LGTM on the dv side

**Comment by tjaychen** [comment]
@timothytrippel can you have a look? does this look okay to you?

**Comment by timothytrippel** [comment]
LGTM, thanks!

**Review by timothytrippel — APPROVED** [review]

**Review by msfschaffner — APPROVED** [review]
LGTM

**Pull request merged**